resin sync/ssh improvements

lekkas opened the pull request:
## New features
* Automatically **parse '.gitignore'** for file inclusions/exclusions from resin sync by default. Skip parsing with `--skip-gitignore`
* **Automatically save options** to `<sourceDirectory>/.resin-sync.yml` after every run
* Support **user-specified destination directories** with `--destination/-d` option
* Implement `--after` option to perform actions local (e.g. cleanup) after resin sync has finished
* Interactive dialog for destination directory, with `/usr/src/app` being the default choice

## Changes
* `resin sync` **`--source/-s`** option is mandatory if a `.resin-sync.yml` file is not found in the current directory
* `resin sync/ssh` now only accept `uuid` as an argument (`appName` has been deprecated)
* Always display interactive device selection dialog when uuid is not passed as an argument

## Fixes
* Disable ControlMaster ssh option (as reported in support)

jviotti commented on lib/actions/sync.coffee:
```diff
@@ -14,6 +14,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ###
 
+
```
I believe this was accidentally added.

lekkas replied:
Yeap, fixed

jviotti commented on lib/utils/helpers.coffee:
```diff
@@ -52,3 +55,24 @@ exports.sudo = (command, message) ->
+# TODO: Use 'config.load()' method from `resin sync` when resin sync gets
+# integrated into resin CLI
+exports.loadConfig = (source) ->
+	configPath = path.join(source, '.resin-sync.yml')
```
We shouldn't hardcode `.resin-sync.yml` here if we call the function `loadConfig()`, since that implies `.resin-sync.yml` is the configuration for the CLI as a whole. I propose either:

- Accepting the configuration file name as an argument.
- Changing the name to `loadSyncConfig()`.
- Moving this function to a private function in `sync.coffee`.

I personally prefer the third one.

lekkas replied:
Good point @jviotti , moved `loadConfig()` as a private function in `sync.coffee.

jviotti commented on lib/actions/sync.coffee:
```diff
@@ -79,6 +91,8 @@ module.exports =
 		,
 	]
 	action: (params, options, done) ->
+		fs = require('fs')
```
Are these imports required?

lekkas replied:
I generally run `resin-lint` to catch unused variables - we use all of the imported modules later on.

jviotti commented:
I didn't check in detail, but I believe some of your changes are missing the JS counter-part, but ignore me if I'm wrong. Also, you can run `npm run doc` to auto generate the documentation at `doc/cli.markdown`; otherwise it'll keep referring to the old command helps.

lekkas commented:
Also added docs (thanks for reminding me, maybe we can add this in the `gulp build` process) and regenerated JS files. I've added the latter in a separate commit at the end, because having them along with the 'legit' coffeescript changes is makes life harder when reviewing the diffs.

lekkas commented:
@jviotti I forgot an important detail - please do not merge/publish this until we have the respective backend changes on production - I will ping you when it's up!

jviotti commented:
Yeah, transpiling is a PITA, I'd love to move away from CoffeeScript. Sure, let me know when everything is ready!

lekkas commented:
Thank you!

Tbh I like the syntactic sugar of CoffeeScript , but I agree that leaving it is probably the most sensible choice for the long term!

alexandrosm commented:
We might move from CS to TS -- so nobody gets what they want :P

-- 

*Alexandros Marinos*

Founder & CEO, Resin.io


+1 206-637-5498

@alexandrosm

On Thu, Jul 21, 2016 at 11:15 AM, Kostas Lekkas <notifications@github.com>
wrote:

> Thank you!
>
> Tbh I like the syntactic sugar of CoffeeScript , but I agree that leaving
> it is probably the most sensible choice for the long term!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/resin-io/resin-cli/pull/367#issuecomment-234337864>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABLUCKIUFIDOOw-sZ81vbAZNtp3QKoPLks5qX7c2gaJpZM4JRuAQ>
> .
>

jviotti commented:
@alexandrosm In any case, should we discuss a way to deal with transpilation in a way that doesn't annoy developers nor users? The current situation is awful. Executing CS or TS directly is an option for internal code, but not for open-source packages we publish to NPM. Adding a pre-publish hook to compile stuff led me to some issues in the past which made me chose to compile and commit the JS myself.

jviotti commented:
Just a comment: I'd love to make sure we get decent linting tools for whatever language we chose. ESLint for JS is magnitudes of times better than things like CoffeeLint. It makes a HUGE difference for me in other projects.

alexandrosm commented:
How does TSLint look on that spectrum?

https://palantir.github.io/tslint/

btw, I love their "develop" link. hint hint for Etcher :)

-- 

*Alexandros Marinos*

Founder & CEO, Resin.io


+1 206-637-5498

@alexandrosm

On Thu, Jul 21, 2016 at 1:58 PM, Juan Cruz Viotti <notifications@github.com>
wrote:

> Just a comment: I'd love to make sure we get decent linting tools for
> whatever language we chose. ESLint for JS is magnitudes of times better
> than things like CoffeeLint. It makes a HUGE difference for me in other
> projects.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/resin-io/resin-cli/pull/367#issuecomment-234381954>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABLUCLF2oVd_KS_QrWqIFKcTVcfEkWJiks5qX91fgaJpZM4JRuAQ>
> .
>

lekkas commented:
@alexandrosm +1 for TS! 😆 

@jviotti What were the issues with adding pre-publishing hooks? 

I agree it'd be better to have transpilation automated. The current situation is a nuisance for the PR submitter and also doubles the review workload for the maintainer - because PR'ed transpiled code *must* be reviewed (or at least verified) as well

jviotti commented:
@lekkas Some I remember from the top of my head:

- I encountered some situations where the CoffeeScript would not compile correctly. The code would compile, but the intention was lost given some edge cases and weird CS usage, and I only noticed after publishing the module. I do a quick sanity review on the compiled JS since then.

- A combination of `.gitignore` and `.npmignore` subtle issues caused the CS to be compiled in the pre-publish hook, but not be included in the final package. It was hard to diagnose what was going on afterwards.

I believe there were other issues/annoyances, but I can't remember anymore.

lekkas commented:
I see, I'm thinking of adding a prepublish hook in `resin sync`, so I'll take note of these issues, thanks!

lekkas commented:
@jviotti this is ready for merging 👍
